File: python/tests/integration/_async/test_mobile_agent_integration.py
"""Integration tests for Mobile Agent functionality."""

import asyncio
import os

import pytest
import pytest_asyncio

from agentbay import AsyncAgentBay
from agentbay import get_logger
from agentbay import CreateSessionParams

from dotenv import load_dotenv

logger = get_logger("mobile-agent-integration-test")
load_dotenv()


@pytest_asyncio.fixture(scope="module")
async def agent_bay():
    """Create an AsyncAgentBay instance."""
    api_key = os.getenv("AGENTBAY_API_KEY")
    if not api_key:
        pytest.skip("AGENTBAY_API_KEY environment variable not set")
    return AsyncAgentBay(api_key=api_key)


@pytest_asyncio.fixture(scope="module")
async def mobile_agent_session(agent_bay):
    """Create a session for mobile agent testing."""
    # Ensure a delay to avoid session creation conflicts
    await asyncio.sleep(3)
    params = CreateSessionParams(
        image_id="imgc-0aae4rgien5oudgb6",
    )
    session_result = await agent_bay.create(params)
    if not session_result.success or not session_result.session:
        # 打印错误信息以便调试
        print(f"\n❌ Failed to create session: {session_result.error_message}")
        logger.error(f"Failed to create session: {session_result.error_message}")
        pytest.skip("Failed to create session")

    session = session_result.session
    # 添加 session_id 打印
    print(f"\n✅ Session created: {session.session_id}")
    logger.info(f"Session created: {session.session_id}")
    yield session

    # Clean up session
    try:
        await session.delete()
        print(f"✅ Session deleted: {session.session_id}")
    except Exception as e:
        logger.info(f"Warning: Error deleting session: {e}")


@pytest.mark.asyncio
async def test_mobile_execute_task_success(mobile_agent_session):
    """Test execute_task basic functionality - non-blocking execution."""
    agent = mobile_agent_session.agent
    task_id = None

    try:
        task = "Open WeChat app"
        logger.info("🚀 Testing execute_task - non-blocking execution")
        
        # Execute task and get task_id immediately
        result = await agent.mobile.execute_task(
            task, max_steps=1, max_step_retries=1
        )
        task_id = result.task_id
        
        # Print results
        print(f"\n{'='*60}")
        print(f"📋 Request ID: {result.request_id}")
        print(f"📋 Task ID: {result.task_id}")
        print(f"📋 Task Status: {result.task_status}")
        print(f"📋 Success: {result.success}")
        print(f"📋 Error Message: {result.error_message}")
        print(f"{'='*60}\n")
        
        # Verify execute_task returned successfully
        assert result.success, f"execute_task failed: {result.error_message}"
        assert result.request_id != "", "Request ID should not be empty"
        assert result.task_id != "", "Task ID should not be empty"
        assert result.error_message == "", f"Error message should be empty: {result.error_message}"
        assert result.task_status == "running", f"Initial status should be 'running', got: {result.task_status}"
        
        logger.info(f"✅ execute_task succeeded: task_id={result.task_id}")
        
        # Poll until task completes to avoid blocking subsequent tests
        max_try_times = 100
        retry_times = 0
        import json

        while retry_times < max_try_times:
            # Call the underlying API directly to get raw response
            raw_result = await agent.mobile.session.call_mcp_tool(
                agent.mobile._get_tool_name("get_status"), {"task_id": task_id}
            )
            
            # Verify the response
            assert raw_result.success, f"get_task_status failed: {raw_result.error_message}"
            
            if raw_result.success and raw_result.data:
                try:
                    content = json.loads(raw_result.data)
                    
                    # Only print content
                    print(f"\n📋 Content (Attempt #{retry_times + 1}):")
                    print(json.dumps(content, indent=2, ensure_ascii=False))
                    print()
                    
                    # Check for error first (task finished or not found)
                    if "error" in content:
                        logger.info(f"✅ Task returned error (task finished or not found): {content.get('error')}")
                        break
                    
                    # Support both taskId (camelCase) and task_id (snake_case)
                    content_task_id = content.get("taskId") or content.get("task_id", task_id)
                    task_status = content.get("status", "unknown")
                    
                    assert content_task_id == task_id, f"Task ID mismatch: expected {task_id}, got {content_task_id}"
                    
                    logger.info(f"⏳ Polling task status: {task_status} (attempt {retry_times + 1})")
                    
                    # If task completed, end polling immediately
                    if task_status == "completed":
                        logger.info(f"✅ Task reached completed status")
                        break
                    
                    if task_status in ["failed", "unsupported"]:
                        logger.info(f"✅ Task reached final status: {task_status}")
                        break
                except json.JSONDecodeError as e:
                    logger.error(f"Failed to parse response: {e}")
                    print(f"📋 Raw Content (parse error): {raw_result.data}")
                    break
            
            retry_times += 1
            await asyncio.sleep(3)
        
        # Verify task completed
        assert retry_times < max_try_times, f"Task did not complete within {max_try_times} attempts"
        logger.info(f"✅ Task completed: status={task_status}")
        
    finally:
        # Final cleanup: terminate task if still running
        if task_id:
            try:
                status = await agent.mobile.get_task_status(task_id)
                if status.task_status == "running":
                    logger.info(f"Cleaning up: terminating task {task_id}")
                    terminate_result = await agent.mobile.terminate_task(task_id)
                    if terminate_result.success:
                        logger.info(f"✅ Task {task_id} terminated successfully")
            except Exception as e:
                logger.debug(f"Could not terminate task {task_id}: {e}")


@pytest.mark.asyncio
async def test_mobile_execute_task_and_wait_success(mobile_agent_session):
    """Test execute_task_and_wait - blocking execution with SDK polling."""
    agent = mobile_agent_session.agent
    task_id = None

    try:
        task = "Open WeChat app"
        logger.info("🚀 Testing execute_task_and_wait - blocking execution")
        
        # Execute task and wait for completion using SDK's built-in polling
        result = await agent.mobile.execute_task_and_wait(
            task,
            max_steps=1,
            max_step_retries=1,
            max_try_times=100  # Only 1 attempt for quick test
        )
        task_id = result.task_id
        
        # Print results
        print(f"\n{'='*60}")
        print(f"📋 Request ID: {result.request_id}")
        print(f"📋 Task ID: {result.task_id}")
        print(f"📋 Task Status: {result.task_status}")
        print(f"📋 Success: {result.success}")
        print(f"📋 Error Message: {result.error_message}")
        if result.task_result:
            print(f"📋 Task Result: {result.task_result}")
        print(f"{'='*60}\n")
        
        # Verify execute_task_and_wait behavior
        assert result.request_id != "", "Request ID should not be empty"
        if result.task_id:
            assert result.task_id != "", "Task ID should not be empty if provided"
        
        logger.info(f"✅ execute_task_and_wait completed: status={result.task_status}")
        
        # If task didn't complete (timeout), ensure it's terminated
        if result.task_status == "running" or (not result.success and task_id):
            logger.info(f"Task still running or failed, ensuring termination...")
            if task_id:
                terminate_result = await agent.mobile.terminate_task(task_id)
                if terminate_result.success:
                    logger.info(f"✅ Task {task_id} terminated successfully")
                else:
                    # Poll until task completes naturally
                    max_try_times = 100
                    retry_times = 0
                    while retry_times < max_try_times:
                        status = await agent.mobile.get_task_status(task_id)
                        if status.task_status in ["completed", "finished", "failed", "cancelled", "unsupported"]:
                            logger.info(f"✅ Task completed: {status.task_status}")
                            break
                        retry_times += 1
                        await asyncio.sleep(3)
        
    finally:
        # Final cleanup: terminate task if still running
        if task_id:
            try:
                status = await agent.mobile.get_task_status(task_id)
                if status.task_status == "running":
                    logger.info(f"Cleaning up: terminating task {task_id}")
                    terminate_result = await agent.mobile.terminate_task(task_id)
                    if terminate_result.success:
                        logger.info(f"✅ Task {task_id} terminated successfully")
            except Exception as e:
                logger.debug(f"Could not terminate task {task_id}: {e}")


@pytest.mark.asyncio
async def test_mobile_get_task_status_success(mobile_agent_session):
    """Test get_task_status - query task status and verify status changes."""
    agent = mobile_agent_session.agent
    task_id = None

    try:
        task = "Open WeChat app"
        logger.info("🚀 Testing get_task_status - status querying")
        
        # First, execute a task
        execute_result = await agent.mobile.execute_task(
            task, max_steps=1, max_step_retries=1
        )
        assert execute_result.success, f"execute_task failed: {execute_result.error_message}"
        task_id = execute_result.task_id
        
        logger.info(f"✅ Task started: task_id={task_id}")
        
        # Query status multiple times to verify status changes, and poll until completion
        max_try_times = 100
        retry_times = 0
        statuses_seen = []
        
        while retry_times < max_try_times:
            status_result = await agent.mobile.get_task_status(task_id)
            
            # Print status
            print(f"\n{'='*60}")
            print(f"📋 Query #{retry_times + 1}")
            print(f"📋 Request ID: {status_result.request_id}")
            print(f"📋 Task ID: {status_result.task_id}")
            print(f"📋 Task Status: {status_result.task_status}")
            print(f"📋 Task Action: {status_result.task_action}")
            print(f"📋 Success: {status_result.success}")
            print(f"📋 Error Message: {status_result.error_message}")
            print(f"{'='*60}\n")
            
            # Verify get_task_status returned successfully
            assert status_result.success, f"get_task_status failed: {status_result.error_message}"
            assert status_result.task_id == task_id, "Task ID should match"
            assert status_result.request_id != "", "Request ID should not be empty"
            
            statuses_seen.append(status_result.task_status)
            logger.info(f"✅ Query #{retry_times + 1}: status={status_result.task_status}, action={status_result.task_action}")
            
            # If task finished, we can stop querying
            if status_result.task_status in ["completed", "finished", "failed", "cancelled", "unsupported"]:
                logger.info(f"✅ Task reached final status: {status_result.task_status}")
                break
            
            retry_times += 1
            await asyncio.sleep(3)
        
        # Verify task completed
        assert retry_times < max_try_times, f"Task did not complete within {max_try_times} attempts"
        assert len(statuses_seen) > 0, "Should have queried status at least once"
        logger.info(f"✅ Statuses seen: {statuses_seen}")
        
    finally:
        # Clean up: terminate task if still running
        if task_id:
            try:
                status = await agent.mobile.get_task_status(task_id)
                if status.task_status == "running":
                    logger.info(f"Cleaning up: terminating task {task_id}")
                    terminate_result = await agent.mobile.terminate_task(task_id)
                    if terminate_result.success:
                        logger.info(f"✅ Task {task_id} terminated successfully")
            except Exception as e:
                logger.debug(f"Could not terminate task {task_id}: {e}")


@pytest.mark.asyncio
async def test_mobile_terminate_task_success(mobile_agent_session):
    """Test terminate_task - terminate a running task."""
    agent = mobile_agent_session.agent
    task_id = None

    try:
        task = "Open WeChat app"
        logger.info("🚀 Testing terminate_task - task termination")
        
        # First, execute a task
        execute_result = await agent.mobile.execute_task(
            task, max_steps=1, max_step_retries=1
        )
        assert execute_result.success, f"execute_task failed: {execute_result.error_message}"
        task_id = execute_result.task_id
        
        logger.info(f"✅ Task started: task_id={task_id}")
        
        # Wait a moment to ensure task is running
        await asyncio.sleep(2)
        
        # Verify task is running before termination
        status_before = await agent.mobile.get_task_status(task_id)
        assert status_before.success, f"get_task_status failed: {status_before.error_message}"
        logger.info(f"📋 Task status before termination: {status_before.task_status}")
        
        # Terminate the task
        terminate_result = await agent.mobile.terminate_task(task_id)
        
        # Print results
        print(f"\n{'='*60}")
        print(f"📋 Request ID: {terminate_result.request_id}")
        print(f"📋 Task ID: {terminate_result.task_id}")
        print(f"📋 Task Status: {terminate_result.task_status}")
        print(f"📋 Success: {terminate_result.success}")
        print(f"📋 Error Message: {terminate_result.error_message}")
        print(f"{'='*60}\n")
        
        # Verify terminate_task succeeded
        assert terminate_result.success, f"terminate_task failed: {terminate_result.error_message}"
        assert terminate_result.request_id != "", "Request ID should not be empty"
        assert terminate_result.task_id == task_id, "Task ID should match"
        
        logger.info(f"✅ terminate_task succeeded: status={terminate_result.task_status}")
        
        # Verify task status after termination - poll until confirmed terminated/completed
        max_try_times = 20
        retry_times = 0
        while retry_times < max_try_times:
            await asyncio.sleep(1)  # Wait a moment for termination to process
            status_after = await agent.mobile.get_task_status(task_id)
            assert status_after.success, f"get_task_status failed: {status_after.error_message}"
            logger.info(f"📋 Task status after termination (attempt {retry_times + 1}): {status_after.task_status}")
            
            # Task should be terminated, finished, or failed
            if status_after.task_status in ["completed", "finished", "failed", "cancelled", "unsupported"]:
                logger.info(f"✅ Task confirmed in final status: {status_after.task_status}")
                break
            
            retry_times += 1
        
        # Verify task is no longer running
        final_status = await agent.mobile.get_task_status(task_id)
        assert final_status.task_status != "running", f"Task should not be running after termination, got: {final_status.task_status}"
        logger.info(f"✅ Task termination confirmed: {final_status.task_status}")
        
    except Exception as e:
        logger.error(f"❌ Test failed: {e}")
        raise
    finally:
        # Final cleanup attempt - ensure task is completed or terminated
        if task_id:
            try:
                status = await agent.mobile.get_task_status(task_id)
                if status.task_status == "running":
                    logger.info(f"Final cleanup: attempting to terminate task {task_id}")
                    terminate_result = await agent.mobile.terminate_task(task_id)
                    if terminate_result.success:
                        logger.info(f"✅ Task {task_id} terminated in cleanup")
                    else:
                        # Poll until task completes naturally
                        max_try_times = 50
                        retry_times = 0
                        while retry_times < max_try_times:
                            status = await agent.mobile.get_task_status(task_id)
                            if status.task_status in ["completed", "finished", "failed", "cancelled", "unsupported"]:
                                logger.info(f"✅ Task completed in cleanup: {status.task_status}")
                                break
                            retry_times += 1
                            await asyncio.sleep(3)
            except Exception as e:
                logger.debug(f"Final cleanup failed for task {task_id}: {e}")

File: java/agentbay/src/test/java/com/aliyun/agentbay/test/TestMobileAgentIntegration.java
package com.aliyun.agentbay.test;

import com.aliyun.agentbay.AgentBay;
import com.aliyun.agentbay.agent.Agent;
import com.aliyun.agentbay.exception.AgentBayException;
import com.aliyun.agentbay.model.ExecutionResult;
import com.aliyun.agentbay.model.QueryResult;
import com.aliyun.agentbay.model.SessionResult;
import com.aliyun.agentbay.session.CreateSessionParams;
import com.aliyun.agentbay.session.Session;
import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;

import static org.junit.Assert.*;

/**
 * Integration tests for Mobile Agent functionality in AgentBay Java SDK.
 * This test class is equivalent to test_mobile_agent_integration.py in Python SDK.
 */
public class TestMobileAgentIntegration {
    private static AgentBay agentBay;
    private static Session session;
    private static Agent agent;

    /**
     * Get API key for testing
     */
    private static String getTestApiKey() {
        String apiKey = System.getenv("AGENTBAY_API_KEY");
        if (apiKey == null || apiKey.trim().isEmpty()) {
            apiKey = "akm-xxx"; // Replace with your actual API key for testing
            System.out.println("Warning: Using default API key. Set AGENTBAY_API_KEY environment variable for production use.");
        }
        return apiKey;
    }

    /**
     * Set up the test environment by creating a session and initializing agent.
     */
    @BeforeClass
    public static void setUp() throws AgentBayException, InterruptedException {
        // Ensure a delay to avoid session creation conflicts
        Thread.sleep(3000);
        
        String apiKey = getTestApiKey();
        agentBay = new AgentBay(apiKey);
        
        // Create a session with mobile_latest image
        CreateSessionParams params = new CreateSessionParams();
        params.setImageId("imgc-0aae4rgien5oudgb6"); // mobile_latest image ID
        
        System.out.println("Creating a new session for mobile agent testing...");
        SessionResult sessionResult = agentBay.create(params);
        
        if (!sessionResult.isSuccess() || sessionResult.getSession() == null) {
            throw new AgentBayException("Failed to create session: " + sessionResult.getErrorMessage());
        }
        
        session = sessionResult.getSession();
        agent = session.getAgent();

        System.out.println("Session created with ID: " + session.getSessionId());
    }

    /**
     * Clean up resources after all tests.
     */
    @AfterClass
    public static void tearDown() {
        System.out.println("Cleaning up: Deleting the session...");
        try {
            if (session != null) {
                agentBay.delete(session, false);
            }
        } catch (Exception e) {
            System.err.println("Warning: Error deleting session: " + e.getMessage());
        }
    }

    @Test
    public void testMobileExecuteTaskSuccess() throws InterruptedException {
        System.out.println("🚀 Test: Mobile execute_task - non-blocking execution");

        String task = "Open WeChat app";
        String taskId = null;
        
        try {
            ExecutionResult result = agent.getMobile().executeTask(task, 1);

            if (!result.isSuccess()) {
                System.out.println("ExecutionResult details:");
                System.out.println("  Success: " + result.isSuccess());
                System.out.println("  Error Message: " + result.getErrorMessage());
                System.out.println("  Request ID: " + result.getRequestId());
                System.out.println("  Task ID: " + result.getTaskId());
                System.out.println("  Task Status: " + result.getTaskStatus());
            }
            assertTrue("Task execution should succeed. Error: " + result.getErrorMessage(), result.isSuccess());
            assertNotNull("Request ID should not be null", result.getRequestId());
            assertFalse("Request ID should not be empty", result.getRequestId().isEmpty());
            assertEquals("Error message should be empty", "", result.getErrorMessage());
            assertNotNull("Task ID should not be null", result.getTaskId());
            assertFalse("Task ID should not be empty", result.getTaskId().isEmpty());
            assertEquals("Initial status should be running", "running", result.getTaskStatus());

            taskId = result.getTaskId();
            System.out.println("✅ Task ID: " + taskId);

            // Poll until task completes to avoid blocking subsequent tests
            String timeoutEnv = System.getenv("AGENT_TASK_TIMEOUT");
            int timeout = (timeoutEnv != null && !timeoutEnv.isEmpty())
                ? Integer.parseInt(timeoutEnv)
                : 300;
            int pollInterval = 3;
            int maxPollAttempts = timeout / pollInterval;
            int retryTimes = 0;
            QueryResult queryResult = null;

            while (retryTimes < maxPollAttempts) {
                queryResult = agent.getMobile().getTaskStatus(taskId);

                assertTrue("Query should succeed", queryResult.isSuccess());
                System.out.println("⏳ Task " + queryResult.getTaskId() + " status: " + queryResult.getTaskStatus());

                String taskStatus = queryResult.getTaskStatus();
                if ("completed".equals(taskStatus) || "failed".equals(taskStatus) || 
                    "cancelled".equals(taskStatus) || "unsupported".equals(taskStatus)) {
                    break;
                }

                retryTimes++;
                Thread.sleep(pollInterval * 1000);
            }

            // Verify the final task status
            assertNotNull("Query result should not be null", queryResult);
            assertTrue("Task should finish within timeout", retryTimes < maxPollAttempts);
            System.out.println("✅ Task completed with status: " + queryResult.getTaskStatus());
        } finally {
            // Clean up: terminate task if still running
            if (taskId != null && !taskId.isEmpty()) {
                try {
                    QueryResult status = agent.getMobile().getTaskStatus(taskId);
                    if (status.isSuccess() && "running".equals(status.getTaskStatus())) {
                        System.out.println("Cleaning up: terminating task " + taskId);
                        ExecutionResult terminateResult = agent.getMobile().terminateTask(taskId);
                        if (terminateResult.isSuccess()) {
                            System.out.println("✅ Task " + taskId + " terminated successfully");
                            Thread.sleep(1000);
                        }
                    }
                } catch (Exception e) {
                    System.out.println("Could not terminate task " + taskId + ": " + e.getMessage());
                }
            }
        }
    }

    @Test
    public void testMobileExecuteTaskAndWaitSuccess() throws InterruptedException {
        System.out.println("🚀 Test: Mobile execute_task_and_wait - blocking execution");

        String task = "Open WeChat app";
        String taskId = null;
        
        try {
            String timeoutEnv = System.getenv("AGENT_TASK_TIMEOUT");
            int timeout = (timeoutEnv != null && !timeoutEnv.isEmpty())
                ? Integer.parseInt(timeoutEnv)
                : 300;

            ExecutionResult result = agent.getMobile().executeTaskAndWait(task, 10, timeout);
            taskId = result.getTaskId();

            if (!result.isSuccess()) {
                System.out.println("ExecutionResult details:");
                System.out.println("  Success: " + result.isSuccess());
                System.out.println("  Error Message: " + result.getErrorMessage());
                System.out.println("  Request ID: " + result.getRequestId());
                System.out.println("  Task ID: " + result.getTaskId());
                System.out.println("  Task Status: " + result.getTaskStatus());
                System.out.println("  Task Result: " + result.getTaskResult());
            }
            assertTrue("Task execution should succeed. Error: " + result.getErrorMessage(), result.isSuccess());
            assertNotNull("Request ID should not be null", result.getRequestId());
            assertFalse("Request ID should not be empty", result.getRequestId().isEmpty());
            assertEquals("Error message should be empty", "", result.getErrorMessage());
            assertNotNull("Task ID should not be null", result.getTaskId());
            assertFalse("Task ID should not be empty", result.getTaskId().isEmpty());
            assertEquals("Task status should be completed", "completed", result.getTaskStatus());

            System.out.println("✅ Task result: " + result.getTaskResult());
        } finally {
            // Clean up: terminate task if still running
            if (taskId != null && !taskId.isEmpty()) {
                try {
                    QueryResult status = agent.getMobile().getTaskStatus(taskId);
                    if (status.isSuccess() && "running".equals(status.getTaskStatus())) {
                        System.out.println("Cleaning up: terminating task " + taskId);
                        ExecutionResult terminateResult = agent.getMobile().terminateTask(taskId);
                        if (terminateResult.isSuccess()) {
                            System.out.println("✅ Task " + taskId + " terminated successfully");
                            Thread.sleep(1000);
                        }
                    }
                } catch (Exception e) {
                    System.out.println("Could not terminate task " + taskId + ": " + e.getMessage());
                }
            }
        }
    }

    @Test
    public void testMobileGetTaskStatusSuccess() throws InterruptedException {
        System.out.println("🚀 Test: Mobile get_task_status");

        String task = "Open WeChat app";
        String taskId = null;
        
        try {
            ExecutionResult executeResult = agent.getMobile().executeTask(task, 1);

            if (!executeResult.isSuccess()) {
                System.out.println("ExecutionResult details:");
                System.out.println("  Success: " + executeResult.isSuccess());
                System.out.println("  Error Message: " + executeResult.getErrorMessage());
                System.out.println("  Request ID: " + executeResult.getRequestId());
                System.out.println("  Task ID: " + executeResult.getTaskId());
                System.out.println("  Task Status: " + executeResult.getTaskStatus());
            }
            assertTrue("Task execution should succeed. Error: " + executeResult.getErrorMessage(), executeResult.isSuccess());
            taskId = executeResult.getTaskId();
            assertNotNull("Task ID should not be null", taskId);
            assertFalse("Task ID should not be empty", taskId.isEmpty());

            // Poll until task completes to avoid blocking subsequent tests
            String timeoutEnv = System.getenv("AGENT_TASK_TIMEOUT");
            int timeout = (timeoutEnv != null && !timeoutEnv.isEmpty())
                ? Integer.parseInt(timeoutEnv)
                : 300;
            int pollInterval = 3;
            int maxPollAttempts = timeout / pollInterval;
            int retryTimes = 0;
            QueryResult queryResult = null;

            while (retryTimes < maxPollAttempts) {
                queryResult = agent.getMobile().getTaskStatus(taskId);

                assertTrue("Query should succeed", queryResult.isSuccess());
                assertNotNull("Request ID should not be null", queryResult.getRequestId());
                assertEquals("Task ID should match", taskId, queryResult.getTaskId());
                assertNotNull("Task status should not be null", queryResult.getTaskStatus());
                assertFalse("Task status should not be empty", queryResult.getTaskStatus().isEmpty());

                System.out.println("⏳ Query #" + (retryTimes + 1) + ": Task status: " + queryResult.getTaskStatus());
                System.out.println("✅ Task action: " + queryResult.getTaskAction());

                String taskStatus = queryResult.getTaskStatus();
                if ("completed".equals(taskStatus) || "failed".equals(taskStatus) || 
                    "cancelled".equals(taskStatus) || "unsupported".equals(taskStatus)) {
                    System.out.println("✅ Task reached final status: " + taskStatus);
                    break;
                }

                retryTimes++;
                Thread.sleep(pollInterval * 1000);
            }

            // Verify task completed
            assertNotNull("Query result should not be null", queryResult);
            assertTrue("Task should finish within timeout", retryTimes < maxPollAttempts);
            System.out.println("✅ Task completed with status: " + queryResult.getTaskStatus());
        } finally {
            // Clean up: terminate task if still running
            if (taskId != null && !taskId.isEmpty()) {
                try {
                    QueryResult status = agent.getMobile().getTaskStatus(taskId);
                    if (status.isSuccess() && "running".equals(status.getTaskStatus())) {
                        System.out.println("Cleaning up: terminating task " + taskId);
                        ExecutionResult terminateResult = agent.getMobile().terminateTask(taskId);
                        if (terminateResult.isSuccess()) {
                            System.out.println("✅ Task " + taskId + " terminated successfully");
                            Thread.sleep(1000);
                        }
                    }
                } catch (Exception e) {
                    System.out.println("Could not terminate task " + taskId + ": " + e.getMessage());
                }
            }
        }
    }

    @Test
    public void testMobileTerminateTaskSuccess() throws InterruptedException {
        System.out.println("🚀 Test: Mobile terminate_task");

        String task = "Open WeChat app";
        String taskId = null;
        
        try {
            ExecutionResult executeResult = agent.getMobile().executeTask(task, 100);

            if (!executeResult.isSuccess()) {
                System.out.println("ExecutionResult details:");
                System.out.println("  Success: " + executeResult.isSuccess());
                System.out.println("  Error Message: " + executeResult.getErrorMessage());
                System.out.println("  Request ID: " + executeResult.getRequestId());
                System.out.println("  Task ID: " + executeResult.getTaskId());
                System.out.println("  Task Status: " + executeResult.getTaskStatus());
            }
            assertTrue("Task execution should succeed. Error: " + executeResult.getErrorMessage(), executeResult.isSuccess());
            taskId = executeResult.getTaskId();
            assertNotNull("Task ID should not be null", taskId);
            assertFalse("Task ID should not be empty", taskId.isEmpty());

            // Wait a bit for task to start
            Thread.sleep(2000);

            ExecutionResult terminateResult = agent.getMobile().terminateTask(taskId);

            assertTrue("Terminate should succeed", terminateResult.isSuccess());
            assertEquals("Task ID should match", taskId, terminateResult.getTaskId());
            assertEquals("Task status should be cancelling", "cancelling", terminateResult.getTaskStatus());

            System.out.println("✅ Task terminated successfully");

            // Verify task is actually terminated by checking status
            String timeoutEnv = System.getenv("AGENT_TASK_TIMEOUT");
            int timeout = (timeoutEnv != null && !timeoutEnv.isEmpty())
                ? Integer.parseInt(timeoutEnv)
                : 300;
            int pollInterval = 3;
            int maxPollAttempts = Math.min(5, timeout / pollInterval); // Limit to 5 attempts for cleanup
            int retryTimes = 0;

            while (retryTimes < maxPollAttempts) {
                QueryResult queryResult = agent.getMobile().getTaskStatus(taskId);
                if (queryResult.isSuccess()) {
                    String taskStatus = queryResult.getTaskStatus();
                    if ("cancelled".equals(taskStatus) || "completed".equals(taskStatus) || 
                        "failed".equals(taskStatus)) {
                        break;
                    }
                }
                retryTimes++;
                Thread.sleep(pollInterval * 1000);
            }
        } finally {
            // Final cleanup: terminate task if still running
            if (taskId != null && !taskId.isEmpty()) {
                try {
                    QueryResult status = agent.getMobile().getTaskStatus(taskId);
                    if (status.isSuccess() && "running".equals(status.getTaskStatus())) {
                        System.out.println("Final cleanup: attempting to terminate task " + taskId);
                        ExecutionResult terminateResult = agent.getMobile().terminateTask(taskId);
                        if (terminateResult.isSuccess()) {
                            System.out.println("✅ Task " + taskId + " terminated in cleanup");
                            Thread.sleep(1000);
                        } else {
                            // Poll until task completes naturally, but with shorter timeout
                            int maxTryTimes = 5;
                            int retryTimes = 0;
                            while (retryTimes < maxTryTimes) {
                                QueryResult queryStatus = agent.getMobile().getTaskStatus(taskId);
                                if (!queryStatus.isSuccess()) {
                                    break;
                                }
                                if ("completed".equals(queryStatus.getTaskStatus()) || 
                                    "failed".equals(queryStatus.getTaskStatus()) || 
                                    "cancelled".equals(queryStatus.getTaskStatus())) {
                                    System.out.println("✅ Task completed in cleanup: " + queryStatus.getTaskStatus());
                                    break;
                                }
                                retryTimes++;
                                Thread.sleep(1000);
                            }
                        }
                    }
                } catch (Exception e) {
                    System.out.println("Final cleanup failed for task " + taskId + ": " + e.getMessage());
                }
            }
        }
    }

    /**
     * Main method to run tests manually (for debugging purposes).
     * In production, use Maven or IDE test runners.
     */
    public static void main(String[] args) {
        System.out.println("=== Running Mobile Agent Integration Tests ===\n");
        
        TestMobileAgentIntegration test = new TestMobileAgentIntegration();
        
        try {
            TestMobileAgentIntegration.setUp();
            
            System.out.println("\n--- Test 1: Execute Task Success ---");
            test.testMobileExecuteTaskSuccess();
            
            System.out.println("\n--- Test 2: Execute Task And Wait Success ---");
            test.testMobileExecuteTaskAndWaitSuccess();
            
            System.out.println("\n--- Test 3: Get Task Status Success ---");
            test.testMobileGetTaskStatusSuccess();
            
            System.out.println("\n--- Test 4: Terminate Task Success ---");
            test.testMobileTerminateTaskSuccess();
            
            TestMobileAgentIntegration.tearDown();
            
            System.out.println("\n=== All Tests Completed ===");
            
        } catch (Exception e) {
            System.err.println("❌ Test failed: " + e.getMessage());
            e.printStackTrace();
        }
    }
}

